### ISSUE TITLE
Rebuilder: EC rebuilds fail when one of the K first fragments is itself damaged

### ISSUE THREAD

**Issue by vdombrovski** [open-io/oio-sds#2119]
##### ISSUE TYPE

 - Bug Report

##### SDS VERSION

```
5.12.0
```

##### CONFIGURATION

```
Default
```

##### OS / ENVIRONMENT

```
NAME="Ubuntu"
VERSION="18.04.4 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.4 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

##### SUMMARY

EC Fragment rebuild fails in some situations when a source chunk is corrupt, even though enough valid sources do exist in the system.

##### STEPS TO REPRODUCE

```bash
# Create an object using EC. In this example, we'll use EC 3 + 5 (at least 3 healthy fragments required to rebuild)
root@sds-node3:~# openio object create rebuild /etc/hosts --policy ECISALC35D1
+-------+------+----------------------------------+--------+
| Name  | Size | Hash                             | Status |
+-------+------+----------------------------------+--------+
| hosts |  227 | 8D7BD73D91C78B7FC5CDFB6959875D2F | Ok     |
+-------+------+----------------------------------+--------+
root@sds-node3:~# openio object locate rebuild hosts
+-----+---------------------------------------------------------------------------------------------+----------------+----------------------------------+
| Pos | Id                                                                                          | Metachunk size | Metachunk hash                   |
+-----+---------------------------------------------------------------------------------------------+----------------+----------------------------------+
| 0.1 | http://100.121.38.152:6202/713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7 |            227 | 8D7BD73D91C78B7FC5CDFB6959875D2F |
| 0.2 | http://100.121.38.153:6202/8C18AC8CB1F6D295666495F2FCBBA4077B0A28C55277793A479D8B5CAD99DD9D |            227 | 8D7BD73D91C78B7FC5CDFB6959875D2F |
| 0.3 | http://100.121.38.153:6200/7530A5191A76B2AC3683EDF91F70BEA1CEEC29F5F260D635DEDE06265A659533 |            227 | 8D7BD73D91C78B7FC5CDFB6959875D2F |
| 0.4 | http://100.121.38.153:6201/CC1D43BA1E79B7C70B4952D144942742CF29E8147D525623CAB2389C514CA02B |            227 | 8D7BD73D91C78B7FC5CDFB6959875D2F |
| 0.5 | http://100.121.38.154:6202/326BE4BA9BD757F8C005EEDE8F2EE2157E77097B0F51C95DFBDF09F1DD1B2D0A |            227 | 8D7BD73D91C78B7FC5CDFB6959875D2F |
| 0.6 | http://100.121.38.152:6201/7FF2EFFF1420D0F01AB091388493FA9EDD429E0B27C0B498D46BBC8BD505952D |            227 | 8D7BD73D91C78B7FC5CDFB6959875D2F |
| 0.7 | http://100.121.38.154:6201/4E384478FE6F3AF4A395F7D50BF4CD0F54F95ACBB9D3C82AE70246326A444AC0 |            227 | 8D7BD73D91C78B7FC5CDFB6959875D2F |
| 0.0 | http://100.121.38.152:6200/1D455F2036C97218247F91143CC363033EBF239CA1EB4F6745046D3D24C8B276 |            227 | 8D7BD73D91C78B7FC5CDFB6959875D2F |
+-----+---------------------------------------------------------------------------------------------+----------------+----------------------------------+

# Let's corrupt one of the fragments (can be caused by bit-rotting for example)
root@sds-node3:~# truncate -s 10 /mnt/data3/OPENIO/rawx-2/326/326BE4BA9BD757F8C005EEDE8F2EE2157E77097B0F51C95DFBDF09F1DD1B2D0A

# Remove 3 other fragments to minimize sources to rebuild
root@sds-node3:~# curl -X DELETE http://100.121.38.153:6201/CC1D43BA1E79B7C70B4952D144942742CF29E8147D525623CAB2389C514CA02B
root@sds-node3:~# curl -X DELETE http://100.121.38.153:6200/7530A5191A76B2AC3683EDF91F70BEA1CEEC29F5F260D635DEDE06265A659533
root@sds-node3:~# curl -X DELETE http://100.121.38.153:6202/8C18AC8CB1F6D295666495F2FCBBA4077B0A28C55277793A479D8B5CAD99DD9D

# Get the CID/ContentID/ChunkID from the chunk we want to rebuild
root@sds-node3:~# curl -I http://100.121.38.152:6202/713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7
HTTP/1.1 200 OK
Date: Wed, 08 Jun 2022 09:34:44 GMT
Server: Apache
Last-Modified: Wed, 08 Jun 2022 09:32:23 GMT
ETag: "5e0ec61f0e8a3"
Accept-Ranges: bytes
Content-Length: 156
X-oio-chunk-meta-container-id: A4833C3A25B7190B28FAB3584D2267AFE37CF9BA21AD4E78CA9E002FB0B46E9B
X-oio-chunk-meta-content-path: hosts
X-oio-chunk-meta-content-id: F1C8F061ECE00500927D222E6805C8E1
X-oio-chunk-meta-content-version: 1654680743626998
X-oio-chunk-meta-content-storage-policy: ECISALC35D1
X-oio-chunk-meta-content-chunk-method: ec/algo=isa_l_rs_cauchy,k=3,m=5
X-oio-chunk-meta-metachunk-size: 227
X-oio-chunk-meta-metachunk-hash: 8D7BD73D91C78B7FC5CDFB6959875D2F
X-oio-chunk-meta-chunk-id: 713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7
X-oio-chunk-meta-chunk-size: 156
X-oio-chunk-meta-chunk-hash: C3C21039709BD4FDAFAF0D99E37C2FF9
X-oio-chunk-meta-chunk-pos: 0.1
X-oio-chunk-meta-oio-version: 4.2
X-oio-chunk-meta-full-path: MY_ACCOUNT/rebuild/hosts/1654680743626998/F1C8F061ECE00500927D222E6805C8E1

# Remove the content
curl -X DELETE http://100.121.38.152:6202/713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7

# Set the chunk ID and try to rebuild the content
 echo "A4833C3A25B7190B28FAB3584D2267AFE37CF9BA21AD4E78CA9E002FB0B46E9B|F1C8F061ECE00500927D222E6805C8E1|713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7" > /tmp/rebuildme

root@sds-node3:~# oio-blob-rebuilder OPENIO --input-file /tmp/rebuildme
2022-06-08 11:35:26.383 133381 7FD65F485370 log INFO START last_report=2022-06-08T11:35:26 0.00s chunks=0 0.00/s bytes=0 0.00B/s errors=0 0.00% start_time=2022-06-08T11:35:26 0.00s total_chunks=0 0.00/s total_bytes=0 0.00B/s total_errors=0 0.00%
2022-06-08 11:35:26.397 133381 7FD65F485370 log ERROR ERROR while rebuilding chunk OPENIO|A4833C3A25B7190B28FAB3584D2267AFE37CF9BA21AD4E78CA9E002FB0B46E9B|F1C8F061ECE00500927D222E6805C8E1|713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7: Invalid fragment payload in ECPyECLibDriver.reconstruct
2022-06-08 11:35:26.397 133381 7FD65F485370 log INFO DONE last_report=2022-06-08T11:35:26 0.01s chunks=1 71.34/s bytes=0 0.00B/s errors=1 100.00% start_time=2022-06-08T11:35:26 0.01s total_chunks=1 71.33/s total_bytes=0 0.00B/s total_errors=1 100.00%
```

##### EXPECTED RESULTS

The rebuild should succeed because:

- 0.0: Healthy
- 0.1: Missing (the chunk we are trying to rebuild)
- 0.2: Missing
- 0.3: Missing
- 0.4: Missing
- 0.5: Corrupt
- 0.6: Healthy
- 0.7: Healthy

We can thus rebuild from healthy positions 0,6,7 

##### ACTUAL RESULTS

The rebuild fails, because we are trying to grab the first N positions regardless of their health, as long as they return a 200 response. The content read from the body is not checked when choosing chunks, as it would require to read it fully. So when the content is truncated at the EC header, we get the following errors:

```
OPENIO|A4833C3A25B7190B28FAB3584D2267AFE37CF9BA21AD4E78CA9E002FB0B46E9B|F1C8F061ECE00500927D222E6805C8E1|713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7: Invalid fragment payload in ECPyECLibDriver.reconstruct

OR

OPENIO|A71D36CD5DCA8F1C05BC6A14ABBC8CD0BD9B4A62D0DAD47BD10BC28B4348C0DE|A6B1234E82DB05000E4CE641C1205C38|6395ED8FA9A10262CA2B202DE6C6D8502B0938AEBC1C04565A37D704B460A8A2 error The rebuilt chunk is not the correct size
```

A fix would involve either randomly sampling healthy fragments to be chosen for rebuild by adding random.samples here:

https://github.com/open-io/oio-sds/blob/8cae246cf777afb8261b58d0690981f5c5339fa1/oio/api/ec.py#L1208

Very dirty PoC below:
```
root@sds-node3:~# diff -u /usr/lib/python2.7/dist-packages/oio/api/ec.py /usr/lib/python2.7/dist-packages/oio/api/ec_old.py
--- /usr/lib/python2.7/dist-packages/oio/api/ec.py      2022-06-08 11:47:29.132060368 +0200
+++ /usr/lib/python2.7/dist-packages/oio/api/ec_old.py  2022-06-08 11:46:43.368275820 +0200
@@ -1169,8 +1169,8 @@
                     'Not enough valid sources to rebuild')
             self.logger.warning(
                 'Use chunk(s) without size information to rebuild a chunk')
-        import random
-        rebuild_iter = self._make_rebuild_iter(random.sample(resps, nb_data))
+
+        rebuild_iter = self._make_rebuild_iter(resps[:nb_data])
         return assumed_chunk_size, rebuild_iter

     def _make_rebuild_iter(self, resps):

# Eventually works when we brute-force it:
root@sds-node3:~# oio-blob-rebuilder OPENIO --input-file /tmp/rebuildme
2022-06-08 11:47:34.533 133744 7F596CF3C370 log INFO START last_report=2022-06-08T11:47:34 0.00s chunks=0 0.00/s bytes=0 0.00B/s errors=0 0.00% start_time=2022-06-08T11:47:34 0.00s total_chunks=0 0.00/s total_bytes=0 0.00B/s total_errors=0 0.00%
2022-06-08 11:47:34.546 133744 7F596CF3C370 log ERROR ERROR while rebuilding chunk OPENIO|A4833C3A25B7190B28FAB3584D2267AFE37CF9BA21AD4E78CA9E002FB0B46E9B|F1C8F061ECE00500927D222E6805C8E1|713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7: Invalid fragment payload in ECPyECLibDriver.reconstruct
2022-06-08 11:47:34.547 133744 7F596CF3C370 log INFO DONE last_report=2022-06-08T11:47:34 0.01s chunks=1 72.68/s bytes=0 0.00B/s errors=1 100.00% start_time=2022-06-08T11:47:34 0.01s total_chunks=1 72.59/s total_bytes=0 0.00B/s total_errors=1 100.00%

root@sds-node3:~# oio-blob-rebuilder OPENIO --input-file /tmp/rebuildme
2022-06-08 11:47:37.104 133747 7FC283A4D370 log INFO START last_report=2022-06-08T11:47:37 0.00s chunks=0 0.00/s bytes=0 0.00B/s errors=0 0.00% start_time=2022-06-08T11:47:37 0.00s total_chunks=0 0.00/s total_bytes=0 0.00B/s total_errors=0 0.00%
2022-06-08 11:47:37.126 133747 7FC283A4D370 log INFO Successful rebuilding for chunk OPENIO|A4833C3A25B7190B28FAB3584D2267AFE37CF9BA21AD4E78CA9E002FB0B46E9B|F1C8F061ECE00500927D222E6805C8E1|713EE12E372CC3ADAFB2EC4589782E173522101DD1AB97EDCF678937A7327DC7
2022-06-08 11:47:37.127 133747 7FC283A4D370 log INFO DONE last_report=2022-06-08T11:47:37 0.02s chunks=1 43.62/s bytes=156 6805.11B/s errors=0 0.00% start_time=2022-06-08T11:47:37 0.02s total_chunks=1 43.60/s total_bytes=156 6802.14B/s total_errors=0 0.00%
```

Another fix would consist in strongly checking fragment health (at least the actual body size) before selecting chunks.

**Comment by fvennetier** [comment]
Hello Vladimir! Thanks for your report. We are working on it.

**Comment by fvennetier** [comment]
Hello. The commits de7d00ad3500b8e9ce7355574fc50cb924d05b50 and e94cc0525d88bbc5cddb83041fbcfcc47a478e75 should fix this problem. They are included in the [5.13.0](https://github.com/open-io/oio-sds/releases/tag/5.13.0) release.
